cpp/src/arrow/array/validate.cc

```diff
@@ -568,6 +568,9 @@ struct ValidateArrayFullImpl {
   }
 
   Status Visit(const DictionaryType& type) {
+    if (!data.dictionary) {
+      return Status::Invalid("Dictionary array has no dictionary");
+    }
```

Comment:
Normally this is already checked in the top-level `ValidateArray`: https://github.com/apache/arrow/blob/a93ce9907c6f7ae70591b786f040d64e0f7d7109/cpp/src/arrow/array/validate.cc#L375-L382